### ISSUE TITLE
💅 `noUnresolvedImport`: multiple categories of false positives

### ISSUE THREAD

**Issue by jtannas** [biomejs/biome#9143]
### Environment information

<details>

```bash

 bun run biome rage --linter
CLI:
  Version:                      2.4.1
  Color support:                true

Platform:
  CPU Architecture:             x86_64
  OS:                           linux

Environment:
  BIOME_LOG_PATH:                    unset
  BIOME_LOG_PREFIX_NAME:             unset
  BIOME_LOG_LEVEL:                   unset
  BIOME_LOG_KIND:                    unset
  BIOME_CONFIG_PATH:                 unset
  BIOME_THREADS:                     unset
  BIOME_WATCHER_KIND:                unset
  BIOME_WATCHER_POLLING_INTERVAL:    unset
  NO_COLOR:                     unset
  TERM:                         xterm-256color
  JS_RUNTIME_VERSION:           v20.19.6
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         bun/1.3.9

Biome Configuration:
  Status:                       Loaded successfully
  Path:                         biome.json
  Formatter enabled:            true
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  false
  HTML full support enabled:    unset

Linter:
  JavaScript enabled:           unset
  JSON enabled:                 unset
  CSS enabled:                  unset
  GraphQL enabled:              unset
  Recommended:                  true
  Enabled rules:
    a11y/noAccessKey
    a11y/noAriaHiddenOnFocusable
    a11y/noAriaUnsupportedElements
    a11y/noAutofocus
    a11y/noDistractingElements
    a11y/noHeaderScope
    a11y/noInteractiveElementToNoninteractiveRole
    a11y/noLabelWithoutControl
    a11y/noNoninteractiveElementInteractions
    a11y/noNoninteractiveElementToInteractiveRole
    a11y/noNoninteractiveTabindex
    a11y/noPositiveTabindex
    a11y/noRedundantAlt
    a11y/noRedundantRoles
    a11y/noStaticElementInteractions
    a11y/noSvgWithoutTitle
    a11y/useAltText
    a11y/useAnchorContent
    a11y/useAriaActivedescendantWithTabindex
    a11y/useAriaPropsForRole
    a11y/useAriaPropsSupportedByRole
    a11y/useButtonType
    a11y/useFocusableInteractive
    a11y/useGenericFontNames
    a11y/useHeadingContent
    a11y/useHtmlLang
    a11y/useIframeTitle
    a11y/useKeyWithClickEvents
    a11y/useKeyWithMouseEvents
    a11y/useMediaCaption
    a11y/useSemanticElements
    a11y/useValidAnchor
    a11y/useValidAriaProps
    a11y/useValidAriaRole
    a11y/useValidAriaValues
    a11y/useValidAutocomplete
    a11y/useValidLang
    complexity/noAdjacentSpacesInRegex
    complexity/noArguments
    complexity/noBannedTypes
    complexity/noCommaOperator
    complexity/noEmptyTypeParameters
    complexity/noExtraBooleanCast
    complexity/noFlatMapIdentity
    complexity/noImportantStyles
    complexity/noStaticOnlyClass
    complexity/noThisInStatic
    complexity/noUselessCatch
    complexity/noUselessConstructor
    complexity/noUselessContinue
    complexity/noUselessEmptyExport
    complexity/noUselessEscapeInRegex
    complexity/noUselessFragments
    complexity/noUselessLabel
    complexity/noUselessLoneBlockStatements
    complexity/noUselessRename
    complexity/noUselessStringRaw
    complexity/noUselessSwitchCase
    complexity/noUselessTernary
    complexity/noUselessThisAlias
    complexity/noUselessTypeConstraint
    complexity/noUselessUndefinedInitialization
    complexity/useArrowFunction
    complexity/useDateNow
    complexity/useFlatMap
    complexity/useIndexOf
    complexity/useLiteralKeys
    complexity/useNumericLiterals
    complexity/useOptionalChain
    complexity/useRegexLiterals
    complexity/useSimpleNumberKeys
    correctness/noChildrenProp
    correctness/noConstAssign
    correctness/noConstantCondition
    correctness/noConstantMathMinMaxClamp
    correctness/noConstructorReturn
    correctness/noEmptyCharacterClassInRegex
    correctness/noEmptyPattern
    correctness/noGlobalDirnameFilename
    correctness/noGlobalObjectCalls
    correctness/noInnerDeclarations
    correctness/noInvalidBuiltinInstantiation
    correctness/noInvalidConstructorSuper
    correctness/noInvalidDirectionInLinearGradient
    correctness/noInvalidGridAreas
    correctness/noInvalidPositionAtImportRule
    correctness/noInvalidUseBeforeDeclaration
    correctness/noMissingVarFunction
    correctness/noNestedComponentDefinitions
    correctness/noNextAsyncClientComponent
    correctness/noNodejsModules
    correctness/noNonoctalDecimalEscape
    correctness/noPrecisionLoss
    correctness/noPrivateImports
    correctness/noProcessGlobal
    correctness/noQwikUseVisibleTask
    correctness/noReactPropAssignments
    correctness/noRenderReturnValue
    correctness/noRestrictedElements
    correctness/noSelfAssign
    correctness/noSetterReturn
    correctness/noSolidDestructuredProps
    correctness/noStringCaseMismatch
    correctness/noSwitchDeclarations
    correctness/noUndeclaredDependencies
    correctness/noUndeclaredVariables
    correctness/noUnknownFunction
    correctness/noUnknownMediaFeatureName
    correctness/noUnknownProperty
    correctness/noUnknownPseudoClass
    correctness/noUnknownPseudoElement
    correctness/noUnknownTypeSelector
    correctness/noUnknownUnit
    correctness/noUnmatchableAnbSelector
    correctness/noUnreachable
    correctness/noUnreachableSuper
    correctness/noUnresolvedImports
    correctness/noUnsafeFinally
    correctness/noUnsafeOptionalChaining
    correctness/noUnusedFunctionParameters
    correctness/noUnusedImports
    correctness/noUnusedLabels
    correctness/noUnusedPrivateClassMembers
    correctness/noUnusedVariables
    correctness/noVoidElementsWithChildren
    correctness/noVoidTypeReturn
    correctness/noVueDataObjectDeclaration
    correctness/noVueDuplicateKeys
    correctness/noVueReservedKeys
    correctness/noVueReservedProps
    correctness/noVueSetupPropsReactivityLoss
    correctness/useExhaustiveDependencies
    correctness/useGraphqlNamedOperations
    correctness/useHookAtTopLevel
    correctness/useImageSize
    correctness/useImportExtensions
    correctness/useIsNan
    correctness/useJsonImportAttributes
    correctness/useJsxKeyInIterable
    correctness/useParseIntRadix
    correctness/useQwikClasslist
    correctness/useQwikMethodUsage
    correctness/useQwikValidLexicalScope
    correctness/useSingleJsDocAsterisk
    correctness/useUniqueElementIds
    correctness/useValidForDirection
    correctness/useValidTypeof
    correctness/useYield
    performance/noAccumulatingSpread
    performance/noAwaitInLoops
    performance/noBarrelFile
    performance/noDelete
    performance/noDynamicNamespaceImportAccess
    performance/noNamespaceImport
    performance/noReExportAll
    performance/noUnwantedPolyfillio
    performance/useGoogleFontPreconnect
    performance/useTopLevelRegex
    security/noBlankTarget
    security/noDangerouslySetInnerHtml
    security/noDangerouslySetInnerHtmlWithChildren
    security/noGlobalEval
    style/noDescendingSpecificity
    style/noNonNullAssertion
    style/useArrayLiterals
    style/useConst
    style/useDeprecatedReason
    style/useExponentiationOperator
    style/useExportType
    style/useImportType
    style/useLiteralEnumMembers
    style/useNodejsImportProtocol
    style/useShorthandFunctionType
    style/useTemplate
    suspicious/noApproximativeNumericConstant
    suspicious/noAssignInExpressions
    suspicious/noAsyncPromiseExecutor
    suspicious/noBiomeFirstException
    suspicious/noCatchAssign
    suspicious/noClassAssign
    suspicious/noCommentText
    suspicious/noCompareNegZero
    suspicious/noConfusingLabels
    suspicious/noConfusingVoidType
    suspicious/noConstEnum
    suspicious/noControlCharactersInRegex
    suspicious/noDebugger
    suspicious/noDocumentCookie
    suspicious/noDoubleEquals
    suspicious/noDuplicateAtImportRules
    suspicious/noDuplicateCase
    suspicious/noDuplicateClassMembers
    suspicious/noDuplicateCustomProperties
    suspicious/noDuplicateElseIf
    suspicious/noDuplicateFields
    suspicious/noDuplicateFontNames
    suspicious/noDuplicateJsxProps
    suspicious/noDuplicateObjectKeys
    suspicious/noDuplicateParameters
    suspicious/noDuplicateProperties
    suspicious/noDuplicateSelectorsKeyframeBlock
    suspicious/noEmptyBlock
    suspicious/noEmptyInterface
    suspicious/noExplicitAny
    suspicious/noExtraNonNullAssertion
    suspicious/noFallthroughSwitchClause
    suspicious/noFunctionAssign
    suspicious/noGlobalAssign
    suspicious/noGlobalIsFinite
    suspicious/noGlobalIsNan
    suspicious/noImplicitAnyLet
    suspicious/noImportAssign
    suspicious/noImportantInKeyframe
    suspicious/noIrregularWhitespace
    suspicious/noLabelVar
    suspicious/noMisleadingCharacterClass
    suspicious/noMisleadingInstantiator
    suspicious/noMisrefactoredShorthandAssign
    suspicious/noNonNullAssertedOptionalChain
    suspicious/noOctalEscape
    suspicious/noPrototypeBuiltins
    suspicious/noQuickfixBiome
    suspicious/noRedeclare
    suspicious/noRedundantUseStrict
    suspicious/noSelfCompare
    suspicious/noShadowRestrictedNames
    suspicious/noShorthandPropertyOverrides
    suspicious/noSparseArray
    suspicious/noSuspiciousSemicolonInJsx
    suspicious/noTemplateCurlyInString
    suspicious/noThenProperty
    suspicious/noTsIgnore
    suspicious/noUnknownAtRules
    suspicious/noUnsafeDeclarationMerging
    suspicious/noUnsafeNegation
    suspicious/noUselessEscapeInString
    suspicious/noUselessRegexBackrefs
    suspicious/noWith
    suspicious/useAdjacentOverloadSignatures
    suspicious/useBiomeIgnoreFolder
    suspicious/useDefaultSwitchClauseLast
    suspicious/useGetterReturn
    suspicious/useGoogleFontDisplay
    suspicious/useIsArray
    suspicious/useIterableCallbackReturn
    suspicious/useNamespaceKeyword

Workspace:
  Open Documents:               0

```

</details>

### Rule name

lint/correctness/noUnresolvedImports

### Playground link

https://github.com/tacticus-tools/tp2/actions/runs/22167827603/job/64098939981?pr=97

### Expected result

I want there to be fewer false positives.

List of false positives encountered
- "The import specifier ______ cannot be resolved: resolved to a Node.js built-in"
  - `node:fs`
  - `node:path`
  - `node:url`
- "The module react has no export named _____"
  - [`useMemo`](https://react.dev/reference/react/useMemo)
  - [`useRef`](https://react.dev/reference/react/useRef)
  - [`useId`](https://react.dev/reference/react/useId)
  - [`useState`](https://react.dev/reference/react/useState)
  - [`useEffect`](https://react.dev/reference/react/useEffect)
- "The module vitest has no export named ______"
  - [`beforeEach`](https://vitest.dev/api/#beforeeach)
  - [`afterEach`](https://vitest.dev/api/#aftereach)
- "The import specifier [lucide-react](https://lucide.dev/) cannot be resolved: found directory without index"
- "The module @tanstack/react-query has no export named [QueryClient](https://tanstack.com/query/latest/docs/reference/QueryClient)."
- "The module @base-ui/react has no export named [Combobox](https://base-ui.com/react/components/combobox).

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by siketyan** [comment]
For the lucide-react case see https://github.com/biomejs/biome/issues/7354

**Comment by Mutantpenguin** [comment]
I get the same with:
* `"node:crypto"`
* `"node:process"`
* `import { MessageActivity, stripMentionsText } from "@microsoft/teams.api";`

**Comment by ematipico** [comment]
Is it possible to have some sort of reproduction? @Mutantpenguin  and @jtannas

**Comment by jtannas** [comment]
> Is it possible to have some sort of reproduction? @Mutantpenguin  and @jtannas 

I was unable to reproduce it in the playground so I linked a branch dedicated to reproducing it.

**Comment by ematipico** [comment]
What branch? What playground? I only see a ci run.

**Comment by jtannas** [comment]
> What branch? What playground? I only see a ci run.

The branch attached to that CI run
https://github.com/tacticus-tools/tp2/pull/97

**Comment by ematipico** [comment]
That doesn't count as a minimal reproduction, because there might be user/project settings that could tarnish the issue. We usually ask users a fresh new, **minimal** reproduction

**Comment by Mutantpenguin** [comment]
1. create and open new folder
2. `npm install @biomejs/biome@latest --save-dev` <- this is `2.4.4` right now
3. `npm install @microsoft/teams.api`
4. `npx @biomejs/biome init`
5. in `biome.json`, add `"correctness": "on"` to linter rules
6. create `app.ts`
  ```ts
  import { MessageActivity } from "@microsoft/teams.api";
  ```
7. run biome `npx @biomejs/biome check .\app.ts`
8. error "The module @microsoft/teams.api has no export named MessageActivity."

Didn't happen before with `2.3.15`.

**Comment by TylerNRobertson** [comment]
Same thing is happening on our end, all our `pnpm` workspaced packages now have this error and didn't on 2.3.15.  We also store our package versions in `pnpm` catalogs and anything stored in there now also appears as a false positive but not on 2.3.15.

Keeping ours on that version for now has been the fix but would be great to upgrade once this is resolved.

**Comment by svchostdotexe** [comment]
also an issue for us with bun imports

**Comment by OctavCrs** [comment]
Also happening on my end

**Comment by ematipico** [comment]
> also an issue for us with bun imports

Ah, not sure we support these ones yet. Maybe open a discussion. I believe we have the same issue with deno. Still, worth opening a discussion to collect use cases

**Comment by TylerNRobertson** [comment]
This still seems to be happening on our end even on 2.4.6

**Comment by ematipico** [comment]
@TylerNRobertson file a new issue please, the cause could be something else

**Comment by tsatsujnr139** [comment]
@TylerNRobertson did you create that new issue I could follow the discussion on, it's the same issue on my end after updating to the latest version

**Comment by TylerNRobertson** [comment]
Sorry no I haven't gotten to it yet @tsatsujnr139 I was planning on making it over the weekend / early next week when I got back to looking at upgradnig our tooling on our end again. If you make one Im happy to also jump in and contribute any findings I have from our logs when I try again too or can share the link here once I make one

**Comment by tsatsujnr139** [comment]
@TylerNRobertson I created the issue here.

https://github.com/biomejs/biome/issues/9478

**Comment by yob-yob** [comment]
Just to report... this one `import { redis } from "bun";` is causing false positive as well...
